sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala

```diff
@@ -257,6 +251,11 @@ case class ExpressionEncoder[T](
    */
   def isSerializedAsStruct: Boolean = objSerializer.dataType.isInstanceOf[StructType]
 
+  /**
+   * Returns true if the type `T` is `Option`.
+   */
+  def isOptionType: Boolean = classOf[Option[_]].isAssignableFrom(clsTag.runtimeClass)
```

Comment:
regarding maintainability, I think it's better to add a 
```
// maybe there is a better naming
def isSerializedAsStructForTopLevel: Boolean = {
  isSerializedAsStruct && isOption
}
```

The benefit is, when people call `isSerializedAsStruct`, they will see this similar method and think carefully which one he should use. Otherwise, people may miss to check `isOptionType` easily.

Reply:
How do we use `isSerializedAsStructForTopLevel`? In all cases now, we only need to check `isSerializedAsStruct && !isOptionType`.